File: sdk/dart/joss_plugin.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

typedef JossMethod = FutureOr<dynamic> Function(List<dynamic> args);

/// joss-rpc-v1 runner. Build with `dart compile exe` for a standalone payload.
Future<void> runJossPlugin(Map<String, JossMethod> methods) async {
  var requestId = '';
  Map<String, dynamic> response;
  try {
    final text = await stdin.transform(utf8.decoder).join();
    final request = jsonDecode(text) as Map<String, dynamic>;
    requestId = request['id']?.toString() ?? '';
    if (request['protocol'] != 'joss-rpc-v1') {
      throw StateError('unsupported protocol');
    }
    final method = request['method']?.toString() ?? '';
    final handler = methods[method];
    if (handler == null) throw ArgumentError('unknown method: $method');
    final args = (request['args'] as List<dynamic>?) ?? const [];
    response = {'id': requestId, 'result': await handler(args)};
  } catch (error) {
    response = {
      'id': requestId,
      'error': {'code': error.runtimeType.toString(), 'message': error.toString()}
    };
  }
  stdout.writeln(jsonEncode(response));
}

File: sdk/php/joss_plugin.php
<?php

declare(strict_types=1);

const JOSS_PLUGIN_PROTOCOL = 'joss-rpc-v1';

/**
 * Runs a JP v2 native sidecar over stdin/stdout.
 *
 * A handler receives positional arguments. A Generator emits streaming chunks
 * and its return value becomes the final result.
 *
 * @param array<string, callable> $methods
 */
function runJossPlugin(array $methods): void
{
    $requestId = '';

    try {
        $raw = stream_get_contents(STDIN);
        if ($raw === false || trim($raw) === '') {
            throw new InvalidArgumentException('empty request');
        }

        $request = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
        if (!is_array($request)) {
            throw new InvalidArgumentException('request must be a JSON object');
        }

        $requestId = isset($request['id']) ? (string) $request['id'] : '';
        if (($request['protocol'] ?? null) !== JOSS_PLUGIN_PROTOCOL) {
            throw new InvalidArgumentException('unsupported protocol');
        }

        $method = isset($request['method']) ? (string) $request['method'] : '';
        if ($method === '' || !isset($methods[$method]) || !is_callable($methods[$method])) {
            throw new BadMethodCallException('unknown method: ' . $method);
        }

        $args = $request['args'] ?? [];
        if (!is_array($args) || !array_is_list($args)) {
            throw new InvalidArgumentException('args must be a JSON array');
        }

        $result = $methods[$method](...$args);
        if ($result instanceof Generator) {
            foreach ($result as $chunk) {
                writeJossFrame(['id' => $requestId, 'event' => 'chunk', 'content' => $chunk]);
            }
            $result = $result->getReturn();
        }

        writeJossFrame(['id' => $requestId, 'result' => $result]);
    } catch (Throwable $error) {
        writeJossFrame([
            'id' => $requestId,
            'error' => [
                'code' => (new ReflectionClass($error))->getShortName(),
                'message' => $error->getMessage(),
            ],
        ]);
    }
}

/** @param array<string, mixed> $frame */
function writeJossFrame(array $frame): void
{
    fwrite(STDOUT, json_encode($frame, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES) . PHP_EOL);
    fflush(STDOUT);
}
